myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java

```diff
@@ -118,4 +131,102 @@ public void flexDownCluster(int numInstancesToScaleDown) {
+     */     
+    public void shutdownResourceManager() {
+        Status driverStatus = driverManager.getDriverStatus();
+        
+        if (Status.DRIVER_RUNNING != driverStatus) {
+            LOGGER.warn("Driver is not running. Status: " + driverStatus);
+        } else {
+            // Stop the driver but leave tasks and executor alone.
+            driverStatus = driverManager.stopDriver(true);
+            LOGGER.info("shutdown....driver status " + driverStatus);
+            terminateManager();
+        }      
+    }
+  
+    /**
+     * Shutdown gracefully means the RM, tasks, and executor(s) are stopped but
+     * the RM can be started right back up, really useful for testing.
+     */   
+    @SuppressWarnings("unchecked")
+    public void shutdownGraceful() {
+       Status driverStatus = driverManager.getDriverStatus();
+        
+        if (Status.DRIVER_RUNNING != driverStatus) {
+            LOGGER.warn(
+                    "Cannot kill tasks, as driver is not running. Status: " + driverStatus);
+        } else {
+          
+            LOGGER.info("shutdown....killing tasks...");
+            killTasks(this.schedulerState.getPendingTaskIds());
+            killTasks(this.schedulerState.getStagingTaskIds());
+            killTasks(this.schedulerState.getActiveTaskIds());        
+    
+            LOGGER.info("shutdown....terminating executors...");
+            killExecutors(this.schedulerState.getExecutorInfo());
+    
+            // Aborting the driver does not stop the executor and kill the tasks.
+            driverStatus = driverManager.abortDriver();
+            LOGGER.info("shutdown....driver status " + driverStatus);
+            terminateManager();
+        }       
+    }
+    
+    private void killTasks(Set<TaskID> tasks) {
+        for (TaskID taskIdToKill : tasks) {
+            Status status = this.driverManager.kill(taskIdToKill);
+            this.schedulerState.removeTask(taskIdToKill);
+            Preconditions.checkState(status == Status.DRIVER_RUNNING);
+        }       
+    }
+    
+    private void killExecutors(Map<Protos.ExecutorID, Protos.SlaveID> executorInfo) {
+        // Send a message to each executor to terminate.
+        for (Entry<ExecutorID, SlaveID> entry : executorInfo.entrySet()) {
+            driverManager.sendFrameworkMessage(entry.getKey(), entry.getValue(), createShutdownCommand());
+        }      
+    }
+    
+    private byte [] createShutdownCommand() {
+        FrameworkUpdate msg = new FrameworkUpdate(FrameworkUpdate.Type.Command, FrameworkUpdate.Operation.Shutdown);
+        return new Gson().toJson(msg).getBytes(Charset.defaultCharset());
+    }
+    
+    private void terminateManager() {
+        // Exit in 2 seconds.
+        new Thread(new Runnable() {
+            public void run() {
+                LOGGER.info("shutdown....exiting RM ");
+
+                try { 
+                    Thread.sleep(2000);
```

Comment:
This was simply to give the system time to kill all the tasks and see them
die on the UI....actually is kind of cool to watch the tasks die but of
course if thats something folks want to see it can be removed.


On Wed, Aug 12, 2015 at 9:19 AM, Ken Sipe <notifications@github.com> wrote:

> In
> myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java
> <https://github.com/mesos/myriad/pull/115#discussion_r36857831>:
>
> > +        }
> > +    }
> > +
> > +    private byte [] createShutdownCommand() {
> > +        FrameworkUpdate msg = new FrameworkUpdate(FrameworkUpdate.Type.Command, FrameworkUpdate.Operation.Shutdown);
> > +        return new Gson().toJson(msg).getBytes(Charset.defaultCharset());
> > +    }
> > +
> > +    private void terminateManager() {
> > +        // Exit in 2 seconds.
> > +        new Thread(new Runnable() {
> > +            public void run() {
> > +                LOGGER.info("shutdown....exiting RM ");
> > +
> > +                try {
> > +                    Thread.sleep(2000);
>
> I'm not a fan of "magic" numbers and I have questions regarding this
> section of code...
> 1. why 2000? or 2 secs? what is the value of that number?
> 2. why is it in it's own thread... if you are just exiting the process
> anyway why spin this thread?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesos/myriad/pull/115/files#r36857831>.
>